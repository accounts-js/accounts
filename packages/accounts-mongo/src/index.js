// @flow

import { get } from 'lodash';
import { encryption } from '@accounts/server';
import type {
  CreateUserType,
  UserObjectType,
  SessionType,
} from '@accounts/common';
import { ObjectID } from 'mongodb';

export type MongoOptionsType = {
  collectionName: string,
  sessionCollectionName: string,
  timestamps: {
    createdAt: string,
    updatedAt: string,
  }
};

export type MongoUserObjectType = {
  username?: string,
  profile?: Object,
  services: {
    password?: {
      bcrypt: string,
    },
  },
  emails?: [{
    address: string,
    verified: boolean,
  }],
};

const toMongoID = (objectId) => {
  if (typeof objectId === 'string') {
    return new ObjectID(objectId);
  }

  return objectId;
};

class Mongo {
  options: MongoOptionsType;
  // TODO definition for mongodb connection object
  db: any;
  collection: any;
  sessionCollection: any;

  constructor(db: any, options: MongoOptionsType) {
    const defaultOptions = {
      collectionName: 'users',
      sessionCollectionName: 'sessions',
      timestamps: {
        createdAt: 'createdAt',
        updatedAt: 'updatedAt',
      },
    };
    this.options = { ...defaultOptions, ...options };
    if (get(db, 'constructor.define.name') !== 'Db') {
      throw new Error('A valid database connection object is required');
    }
    this.db = db;
    this.collection = this.db.collection(this.options.collectionName);
    this.sessionCollection = this.db.collection(this.options.sessionCollectionName);
  }

  async setupIndexes(): Promise<void> {
    await this.collection.createIndex('username', { unique: 1, sparse: 1 });
    await this.collection.createIndex('emails.address', { unique: 1, sparse: 1 });
  }

  async createUser(options: CreateUserType): Promise<string> {
    const user: MongoUserObjectType = {
      services: {},
      profile: {},
      [this.options.timestamps.createdAt]: Date.now(),
      [this.options.timestamps.updatedAt]: Date.now(),
    };
    if (options.password) {
      user.services.password = { bcrypt: await encryption.hashPassword(options.password) };
    }
    if (options.username) {
      user.username = options.username;
    }
    if (options.email) {
      user.emails = [{ address: options.email.toLowerCase(), verified: false }];
    }
    if (options.profile) {
      user.profile = options.profile;
    }
    const ret = await this.collection.insertOne(user);
    return ret.ops[0]._id;
  }

  async findUserById(userId: string): Promise<?UserObjectType> {
    const user = await this.collection.findOne({ _id: toMongoID(userId) });
    if (user) {
      user.id = user._id;
    }
    return user;
  }

  async findUserByEmail(email: string): Promise<?UserObjectType> {
    const user = await this.collection.findOne({ 'emails.address': email.toLowerCase() });
    if (user) {
      user.id = user._id;
    }
    return user;
  }

  async findUserByUsername(username: string): Promise<?UserObjectType> {
    const user = await this.collection.findOne({ username });
    if (user) {
      user.id = user._id;
    }
    return user;
  }

  async findPasswordHash(userId: string): Promise<?string> {
    const user = await this.findUserById(toMongoID(userId));
    if (user) {
      return user.services.password.bcrypt;
    }
    return null;
  }

  async addEmail(userId: string, newEmail: string, verified: boolean): Promise<void> {
    const ret = await this.collection.update({ _id: toMongoID(userId) }, {
      $addToSet: {
        emails: {
          address: newEmail.toLowerCase(),
          verified,
        },
      },
      $set: { [this.options.timestamps.updatedAt]: Date.now() },
    });
    if (ret.result.nModified === 0) {
      throw new Error('User not found');
    }
  }

  async removeEmail(userId: string, email: string): Promise<void> {
    const ret = await this.collection.update({ _id: toMongoID(userId) }, {
      $pull: { emails: { address: email.toLowerCase() } },
      $set: { [this.options.timestamps.updatedAt]: Date.now() },
    });
    if (ret.result.nModified === 0) {
      throw new Error('User not found');
    }
  }

  async setUsername(userId: string, newUsername: string): Promise<void> {
    const ret = await this.collection.update({ _id: toMongoID(userId) }, {
      $set: {
        username: newUsername,
        [this.options.timestamps.updatedAt]: Date.now(),
      },
    });
    if (ret.result.nModified === 0) {
      throw new Error('User not found');
    }
  }

  async setPasssword(userId: string, newPassword: string): Promise<void> {
    const ret = await this.collection.update({ _id: toMongoID(userId) }, {
      $set: {
        'services.password.bcrypt': await encryption.hashPassword(newPassword),
        [this.options.timestamps.updatedAt]: Date.now(),
      },
    });
    if (ret.result.nModified === 0) {
      throw new Error('User not found');
    }
  }

  async setProfile(userId: string, profile: Object): Promise<Object> {
    await this.collection.update({ _id: toMongoID(userId) }, {
      $set: {
        profile,
        [this.options.timestamps.updatedAt]: Date.now(),
      },
    });
    const user = await this.findUserById(userId);
    return user && user.profile;
  }

  async createSession(userId: string, ip: string, userAgent: string): Promise<string> {
    const ret = await this.sessionCollection.insertOne({
      userId,
      userAgent,
      ip,
      valid: true,
      [this.options.timestamps.createdAt]: Date.now(),
      [this.options.timestamps.updatedAt]: Date.now(),
    });
    return ret.ops[0]._id;
  }

  async updateSession(sessionId: string, ip: string, userAgent: string): Promise<void> {
    await this.sessionCollection.update({ _id: toMongoID(sessionId) }, {
      $set: {
        ip,
        userAgent,
        [this.options.timestamps.updatedAt]: Date.now(),
      },
    });
  }

  async invalidateSession(sessionId: string): Promise<void> {
    await this.sessionCollection.update({ _id: toMongoID(sessionId) }, {
      $set: {
        valid: false,
        [this.options.timestamps.updatedAt]: Date.now(),
      },
    });
  }

  findSessionById(sessionId: string): Promise<?SessionType> {
    return this.sessionCollection.findOne({ _id: toMongoID(sessionId) });
  }
}

export default Mongo;
