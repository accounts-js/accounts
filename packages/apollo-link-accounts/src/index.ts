import { setContext } from 'apollo-link-context';
import { ApolloLink } from 'apollo-link';
import { AccountsClient } from '@accounts/client';

export const accountsLink = (accountsClient: AccountsClient): ApolloLink => {
  return setContext(async (_, { headers }) => {
    const tokens = await accountsClient.refreshSession();
    return {
      headers: {
        ...headers,
        'accounts-access-token': tokens ? tokens.accessToken : null,
      },
    };
  });
};
