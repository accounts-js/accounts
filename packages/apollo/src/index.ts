import { accountsLink } from '@accounts/apollo-link';
import { AccountsClientOptions, AccountsClient } from '@accounts/client';
import AccountsGraphQLClient from '@accounts/graphql-client';
import { ApolloClient } from 'apollo-client';
import { ApolloLink } from 'apollo-link';
import { onError } from 'apollo-link-error';
import { HttpLink } from 'apollo-link-http';
import { InMemoryCache } from 'apollo-cache-inmemory';

export interface AccountsApolloOptions extends AccountsClientOptions {
  uri: string;
}

export const accountsApollo = (options: AccountsApolloOptions) => {
  const accountsApolloClient = new ApolloClient({
    cache: new InMemoryCache(),
    link: ApolloLink.from([
      onError(({ graphQLErrors, networkError }) => {
        if (graphQLErrors) {
          graphQLErrors.map(({ message, locations, path }) =>
            console.log(
              `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
            )
          );
        }
        if (networkError) {
          console.log(`[Network error]: ${networkError}`);
        }
      }),
      new HttpLink({
        uri: options.uri,
      }),
    ]),
  });

  const accountsClient = new AccountsClient(
    options,
    new AccountsGraphQLClient({
      graphQLClient: accountsApolloClient,
    })
  );

  let accountsPassword;
  if (require.resolve('@accounts/client-password')) {
    const AccountsClientPassword = require('@accounts/client-password').AccountsClientPassword; // tslint:disable-line no-implicit-dependencies
    accountsPassword = new AccountsClientPassword(accountsClient);
  }

  return {
    accountsClient,
    accountsApolloClient,
    accountsLink: accountsLink(accountsClient),
    accountsPassword,
  };
};

export default accountsApollo;
