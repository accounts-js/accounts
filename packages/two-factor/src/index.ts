export { TwoFactor } from './two-factor';
export { AccountsTwoFactorOptions } from './types';
export { getUserTwoFactorService } from './utils';
