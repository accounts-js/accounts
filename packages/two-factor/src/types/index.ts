export { AccountsTwoFactorOptions } from './options';
export { TwoFactorService } from './two-factor-service';
export { ErrorMessages } from './error-messages';
