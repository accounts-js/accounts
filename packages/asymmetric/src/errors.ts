import { ErrorMessages } from './types';

export const errors: ErrorMessages = {
  userNotFound: 'User not found',
  verificationFailed: 'Failed to verify signature',
};
