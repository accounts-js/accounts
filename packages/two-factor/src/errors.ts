import { ErrorMessages } from './types';

export const errors: ErrorMessages = {
  userNotFound: 'User not found',
  codeDidNotMatch: `2FA code didn't match`,
  userTwoFactorNotSet: `2FA not set`,
  userTwoFactorAlreadySet: `2FA already set`,
  codeRequired: '2FA code required',
};
