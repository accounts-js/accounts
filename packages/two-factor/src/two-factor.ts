import * as speakeasy from 'speakeasy';
import { User, DatabaseInterface } from '@accounts/types';
import { errors } from './errors';
import { AccountsTwoFactorOptions } from './types';
import { getUserTwoFactorService } from './utils';

const defaultOptions = {
  secretLength: 20,
  window: 0,
  errors,
};

export class TwoFactor {
  private options: AccountsTwoFactorOptions & typeof defaultOptions;
  private db!: DatabaseInterface;
  private serviceName = 'two-factor';

  constructor(options: AccountsTwoFactorOptions = {}) {
    this.options = { ...defaultOptions, ...options };
  }

  /**
   * Set two factor store
   */
  public setStore(store: DatabaseInterface): void {
    this.db = store;
  }

  /**
   * Authenticate a user with a 2fa code
   */
  public async authenticate(user: User, code: string): Promise<void> {
    if (!code) {
      throw new Error(this.options.errors.codeRequired);
    }

    const twoFactorService = getUserTwoFactorService(user);
    // If user does not have 2fa set return error
    if (!twoFactorService) {
      throw new Error(this.options.errors.userTwoFactorNotSet);
    }
    if (
      !speakeasy.totp.verify({
        secret: twoFactorService.secret.base32,
        encoding: 'base32',
        token: code,
        window: this.options.window,
      })
    ) {
      throw new Error(this.options.errors.codeDidNotMatch);
    }
  }

  /**
   * Generate a new two factor secret
   */
  public getNewAuthSecret(): speakeasy.Key {
    return speakeasy.generateSecret({
      length: this.options.secretLength,
      name: this.options.appName,
    });
  }

  /**
   * Verify the code is correct
   * Add the code to the user profile
   * Throw if user already have 2fa enabled
   */
  public async set(userId: string, secret: speakeasy.Key, code: string): Promise<void> {
    if (!code) {
      throw new Error(this.options.errors.codeRequired);
    }

    const user = await this.db.findUserById(userId);
    if (!user) {
      throw new Error(this.options.errors.userNotFound);
    }
    let twoFactorService = getUserTwoFactorService(user);
    // If user already have 2fa return error
    if (twoFactorService) {
      throw new Error(this.options.errors.userTwoFactorAlreadySet);
    }

    if (
      speakeasy.totp.verify({
        secret: secret.base32,
        encoding: 'base32',
        token: code,
        window: this.options.window,
      })
    ) {
      twoFactorService = {
        secret,
      };
      await this.db.setService(userId, this.serviceName, twoFactorService);
    } else {
      throw new Error(this.options.errors.codeDidNotMatch);
    }
  }

  /**
   * Remove two factor for a user
   */
  public async unset(userId: string, code: string): Promise<void> {
    if (!code) {
      throw new Error(this.options.errors.codeRequired);
    }

    const user = await this.db.findUserById(userId);
    if (!user) {
      throw new Error(this.options.errors.userNotFound);
    }
    const twoFactorService = getUserTwoFactorService(user);
    // If user does not have 2fa set return error
    if (!twoFactorService) {
      throw new Error(this.options.errors.userTwoFactorNotSet);
    }
    if (
      speakeasy.totp.verify({
        secret: twoFactorService.secret.base32,
        encoding: 'base32',
        token: code,
        window: this.options.window,
      })
    ) {
      this.db.unsetService(userId, this.serviceName);
    } else {
      throw new Error(this.options.errors.codeDidNotMatch);
    }
  }
}
