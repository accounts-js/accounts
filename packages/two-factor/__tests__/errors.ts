import { errors } from '../src/errors';

describe('errors', () => {
  it('should have named export errors', () => {
    expect(errors).toMatchSnapshot();
  });
});
