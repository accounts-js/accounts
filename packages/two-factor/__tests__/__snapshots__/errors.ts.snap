// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errors should have named export errors 1`] = `
Object {
  "codeDidNotMatch": "2FA code didn't match",
  "codeRequired": "2FA code required",
  "userNotFound": "User not found",
  "userTwoFactorAlreadySet": "2FA already set",
  "userTwoFactorNotSet": "2FA not set",
}
`;
